File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isValidBST(root) {
        if (!root) true;

        return this.isValidBSTRec(root.left, -Number.MAX_VALUE, root.val)
            && this.isValidBSTRec(root.right, root.val, Number.MAX_VALUE);
    }

    isValidBSTRec(root, min, max) {
        if (!root) return true;

        return root.val > min && root.val < max && this.isValidBSTRec(root.left, min, root.val)
            && this.isValidBSTRec(root.right, root.val, max);
    }
}

File: Data Structures & Algorithms/invert-a-binary-tree/submission-4.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) return null;

        var queue = new ArrayDeque<TreeNode>();
        queue.addLast(root);
        while (!queue.isEmpty()) {
            int n = queue.size();

            for (int i = 0; i < n;  i++) {
                var node =  queue.pollFirst();
                if (node.left != null)  queue.addLast(node.left);
                if (node.right != null)  queue.addLast(node.right);

                var temp  = node.left;
                node.left = node.right;
                node.right = temp;
            }
        }

        return root;
    }
}
